packages/ui/components/form/selectimproved/components/Options.tsx

```diff
@@ -0,0 +1,109 @@
+import React, { useCallback, useContext, useEffect, useMemo, useState } from "react";
+
+import { useKeyPress } from "@calcom/lib/hooks/useKeyPress";
+
+import { Label } from "../../inputs/Label";
+import Item from "./Item";
+import { SelectContext } from "./SelectProvider";
+import { Option } from "./type";
+
+interface OptionsProps<T extends Option> {
+  list: T[];
+  inputValue: string;
+  isMultiple: boolean;
+  selected: T | T[] | null;
+  searchBoxRef: React.RefObject<HTMLInputElement>;
+}
+
+type FlatternedOption = Option & { current: number; groupedIndex?: number };
+
+const flatternOptions = (options: Option[], groupCount?: number): FlatternedOption[] => {
+  return options.reduce((acc, option, current) => {
+    if (option.options) {
+      return [...acc, ...flatternOptions(option.options, current + (groupCount || 0))];
+    }
+    return [...acc, { ...option, current, groupedIndex: groupCount }];
+  }, [] as FlatternedOption[]);
+};
+
+function Options<T extends Option>({ list, inputValue, searchBoxRef }: OptionsProps<T>) {
+  const { classNames, handleValueChange } = useContext(SelectContext);
+  const [keyboardFocus, setKeyboardFocus] = useState(-1);
+  const downPress = useKeyPress("ArrowDown", searchBoxRef);
+  const upPress = useKeyPress("ArrowUp", searchBoxRef);
+  const enterPress = useKeyPress("Enter", searchBoxRef);
+
+  const flatternedList = useMemo(() => flatternOptions(list), [list]);
+
+  const totalOptionsLength = useMemo(() => {
+    return flatternedList.length;
+  }, [flatternedList]);
```

Comment:
Calculate the length of this - needed to get the min and max bounds of the array for keyboard navigation